### PR TITLE
Migrate New-CRsaKeyPair from Carbon and create signing certificates

### PR DESCRIPTION
**Pull request by splatteredbits**
We needed to update New-CRsaKeyPair to support generating signing certificates. so migrated `New-CRsaKeyPair` from Carbon, then added signing key support.

### CODE REVIEW

**Comment by splatteredbits** [comment]
OK. Thjs is ready for review. Finally got the tests to pass.

File: Carbon.Cryptography/Functions/New-CRsaKeyPair.ps1 [hunk 1]
```diff
@@ -0,0 +1,375 @@
+
+function New-CRsaKeyPair
+{
+    <#
+    .SYNOPSIS
+    Generates a public/private RSA key pair.
+
+    .DESCRIPTION
+    The `New-CRsaKeyPair` function uses the `certreq.exe` program to generate an RSA public/private key pair suitable
```
Comment (DecoyJoe) [inline Carbon.Cryptography/Functions/New-CRsaKeyPair.ps1:9]: How should we indicate this function is Windows only? Docs? Write an error when not `$IsWindows`?

File: Carbon.Cryptography/Functions/New-CRsaKeyPair.ps1 [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+    .DESCRIPTION
+    The `New-CRsaKeyPair` function uses the `certreq.exe` program to generate an RSA public/private key pair suitable
+    for use in encrypting/decrypting CMS messages, credentials in DSC resources, etc.
+
+    Pass the subject to the `Subject` parameter (it must begin with `CN=`) and the paths where you want the public and
+    private keys saved to the `PublicKeyPath` and `PrivateKeyPath` parameters, respectively. `New-CRsaKeyPair` creates
+    a temporary .inf file and passes it to the `certreq.exe` program. You will be prompted for a password, unless you
+    pass a password to the `Password` parameter.
+    
+    By default, a key pair with no key usages or enhanced key usages is generated that is 4096 bits in length, usages
```
Comment (DecoyJoe) [inline Carbon.Cryptography/Functions/New-CRsaKeyPair.ps1:17]: Typo: `usages` should be `uses`

File: Carbon.Cryptography/Functions/New-CRsaKeyPair.ps1 [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+    `SHA512` as the signature/hash algorithm, and is valid until December 31st, 9999. An object with
+    `[IO.FileInfo] PublicKeyFile` and `[IO.FileInfo] PrivateKeyFile` properties is returned.
+
+    You can change the key's length, algorithm, and expiration data with the `Length`, `Algorithm`, and `ValidTo`
+    parameters. You can set the key pair's usages with the `KeyUsage` parameter. Valid usages this function supports
+    are `ClientAuthentication`, `CodeSigning`, `DocumentEncryption`, `DocumentSigning`, and `ServerAuthentication`.
+
+    To create  key pair with no password on the private key, pass `$null` to the `Password` parameter.
```
Comment (DecoyJoe) [inline Carbon.Cryptography/Functions/New-CRsaKeyPair.ps1:25]: Typo: Missing `a` between `create   key`

File: Carbon.Cryptography/Functions/New-CRsaKeyPair.ps1 [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+
+    If the destination files already exist, you'll get an error and no keys will be generated. Use the `Force` switch to
+    overwrite any existing files.
+
+    The `certreq.exe` command stores the private key in the current user's `My` certificate store. This function exports
+    that private key to a file and removes it from the current user's `My` store. The private key is protected with the
+    password provided via the `-Password` parameter. If you don't provide a password, you will be prompted for one. To
+    not protect the private key with a password, pass `$null` as the value of the `-Password` parameter.
```
Comment (DecoyJoe) [inline Carbon.Cryptography/Functions/New-CRsaKeyPair.ps1:33]: This is said twice, on line 25 and here.

File: Carbon.Cryptography/Functions/New-CRsaKeyPair.ps1 [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+    .LINK
+    Install-CCertificate
+
+    .EXAMPLE
+    New-CRsaKeyPair -Subject 'CN=MyName' -PublicKeyFile 'MyName.cer' -PrivateKeyFile 'MyName.pfx' -Password $secureString
+
+    Demonstrates the minimal parameters needed to generate a key pair. The key will use a sha512 signing algorithm, have
+    a length of 4096 bits, and expire on `12/31/9999`. The public key will be saved in the current directory as
+    `MyName.cer`. The private key will be saved to the current directory as `MyName.pfx` and protected with password in
+    `$secureString`. The key pair will no usages.
```
Comment (DecoyJoe) [inline Carbon.Cryptography/Functions/New-CRsaKeyPair.ps1:50]: Missing a word: `The key pair will no usages.`

File: Carbon.Cryptography/Functions/New-CRsaKeyPair.ps1 (hunk 1 above)
Comment (splatteredbits) [reply Carbon.Cryptography/Functions/New-CRsaKeyPair.ps1:9]: Line 213:
```powershell
        $certReqPath = Get-Command -Name 'certreq.exe' | Select-Object -ExpandProperty 'Path'
        if( -not $certReqPath )
        {
            return
        }
```

Note that `Get-Command` doesn't have an `ErrorAction` parameter, so if the `certreq.exe` command doesn't exist, the user will get an error message that `certreq.exe` doesn't exist.

Comment (splatteredbits) [reply Carbon.Cryptography/Functions/New-CRsaKeyPair.ps1:9]: I changed it to this:

```powershell
        $certReqPath = Get-Command -Name 'certreq.exe' -ErrorAction Ignore | Select-Object -ExpandProperty 'Path'
        if( -not $certReqPath )
        {
            'Command "certreq.exe" does not exist. This is a Windows-only command. If you''re on Windows, make sure ' +
            '"C:\Windows\System32" is part of your "Path" environment variable.' |
                Write-Error -ErrorAction $ErrorActionPreference
            return
        }
```

**Pull request merged**